File: typescript/packages/core/src/utils/addParametersToDescription.ts
import type { z } from "zod";

export function addParametersToDescription(description: string, schema: z.ZodTypeAny): string {
    let paramLines: string[] = [];

    if (schema._def?.typeName === "ZodObject") {
        const shape = schema._def.shape();
        if (shape) {
            paramLines = Object.entries<z.ZodTypeAny>(shape).map(([key, value]) => {
                const isOptional = value.isOptional();
                const paramDescription = value.description || "";
                const typeStr = getTypeString(value);
                return `- ${key}${isOptional ? " (optional)" : ""} (${typeStr}): ${paramDescription}`;
            });
        }
    }

    return `${description}\n${paramLines.join("\n")}`;
}

function getTypeString(schema: z.ZodTypeAny): string {
    const typeName = schema._def?.typeName;

    switch (typeName) {
        case "ZodOptional":
            return getTypeString((schema as z.ZodOptional<z.ZodTypeAny>).unwrap());
        case "ZodString":
            return "string";
        case "ZodNumber":
            return "number";
        case "ZodBoolean":
            return "boolean";
        case "ZodArray":
            return "array";
        case "ZodObject":
            return "object";
        default:
            return "unknown";
    }
}

File: python/src/goat-sdk/goat/utils/add_parameters_to_description.py
from zon import Zon, ZonOptional, ZonRecord, ZonString, ZonNumber, ZonBoolean, ZonList, ZonContainer

def add_parameters_to_description(description: str, schema: ZonRecord) -> str:
    """Add parameter descriptions to a tool description."""
    param_lines: list[str] = []
    
    if isinstance(schema, ZonRecord):
        shape = schema.shape
        if shape:
            for key, value in shape.items():
                is_optional = isinstance(value, ZonOptional)
                param_description = value.description or "" # pyright: ignore
                type_str = get_type_string(value)
                
                param_line = f"- {key}"
                if is_optional:
                    param_line += " (optional)"
                param_line += f" ({type_str}): {param_description}"
                
                param_lines.append(param_line)
    
    return f"{description}\n{"\n".join(param_lines)}" 

def get_type_string(field_type: Zon) -> str:
    """Get a string representation of a type."""
    if isinstance(field_type, ZonOptional):
        return get_type_string(field_type.unwrap())
    elif isinstance(field_type, ZonString):
        return "string"
    elif isinstance(field_type, ZonNumber):
        return "number"
    elif isinstance(field_type, ZonBoolean):
        return "boolean"
    elif isinstance(field_type, ZonList):
        return "array"
    elif isinstance(field_type, ZonContainer):
        return "object"
    else:
        return "unknown"
